tsdb/head.go

```diff
@@ -934,18 +941,89 @@ func (h *Head) Truncate(mint int64) (err error) {
 	return h.truncateWAL(mint)
 }
 
+// AcquireQuerierRLock waits till there is not Head truncation happening in parallel
+// and acquires a lock to prevent parallel truncation.
+// * return true means it acquired the lock and it is mandatory to h.ReleaseQuerierLock()
+//   after the query.
+// * return false means the lock was not acquired and it's safe to query.
+func (h *Head) AcquireQuerierRLock(mint, maxt int64) bool {
+	if h.memTruncationInProcess.Load() {
+		// We assume that truncation is not called too often.
+		// So if truncation is in progress, it is safe to check
+		// the last truncation time as best effort to avoid taking the lock if
+		// the query does not overlap the truncation.
+		if h.lastMemoryTruncationStartTime.Load() > maxt || mint > h.lastMemoryTruncationEndTime.Load() {
+			// Does not overlap with the truncation range.
+			return false
+		}
+	}
+
+	// If we reach here, it means either the truncation is not in progress or
+	// we overlap with the truncation range in progress.
+	h.memTruncateMtx.RLock()
+	return true
+}
+
+func (h *Head) ReleaseQuerierRLock() {
+	h.memTruncateMtx.RUnlock()
+}
+
+func (h *Head) AcquireQuerierLock() {
+	h.memTruncateMtx.Lock()
+}
+
+func (h *Head) ReleaseQuerierLock() {
+	h.memTruncateMtx.Unlock()
+}
+
+func (h *Head) WaitForPendingReadersInTimeRange(mint, maxt int64) {
```

Comment:
This seems like an issue to me as it would block for up to 2 minutes, while we do not execute new queries. Could we think about an alternative?